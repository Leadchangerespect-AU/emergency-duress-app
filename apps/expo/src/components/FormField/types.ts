export interface FormFieldProps {
  label: string;
  isRequired?: boolean;
  isDisabled?: boolean;
  errors: (undefined | false | null | string)[];
}
