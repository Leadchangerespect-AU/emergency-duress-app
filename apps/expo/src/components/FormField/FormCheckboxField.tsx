import {
  FormControl,
  FormControlError,
  FormControlErrorIcon,
  FormControlErrorText,
  FormControlLabel,
} from "@/components/FormControl";
import Checkbox, { CheckboxProps } from "../Checkbox";
import { FormFieldProps } from "./types";

export const FormCheckboxField = ({
  isRequired,
  errors,
  isDisabled,
  children,
  ...props
}: Omit<FormFieldProps, "label"> & CheckboxProps) => {
  const hasErrors = errors.length > 0;

  return (
    <FormControl
      isInvalid={hasErrors}
      isRequired={isRequired}
      isDisabled={isDisabled}
    >
      <Checkbox {...props}>{children}</Checkbox>
      {hasErrors && (
        <FormControlError>
          <FormControlErrorIcon />
          <FormControlErrorText>{errors.join(", ")}</FormControlErrorText>
        </FormControlError>
      )}
    </FormControl>
  );
};
