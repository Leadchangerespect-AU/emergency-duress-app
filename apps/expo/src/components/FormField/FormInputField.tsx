import { useId } from "react";
import { Platform, TextInputProps } from "react-native";

import {
  FormControl,
  FormControlError,
  FormControlErrorIcon,
  FormControlErrorText,
  FormControlLabel,
} from "@/components/FormControl";
import Input from "@/components/Input";
import { FormFieldProps } from "./types";

export const FormInputField = ({
  label,
  isRequired,
  isDisabled,
  errors,
  ...props
}: FormFieldProps & TextInputProps) => {
  const hasErrors = errors.length > 0;

  const labelId = useId();
  const inputId = useId();
  return (
    <FormControl
      isInvalid={hasErrors}
      isRequired={isRequired}
      isDisabled={isDisabled}
    >
      <FormControlLabel
        id={labelId}
        {...Platform.select({
          web: {
            htmlFor: inputId,
          },
        })}
      >
        {label}
      </FormControlLabel>
      <Input
        {...props}
        aria-labelledby={labelId}
        accessibilityLabel={label}
        id={inputId}
      />
      {hasErrors && (
        <FormControlError>
          <FormControlErrorIcon />
          <FormControlErrorText>{errors.join(", ")}</FormControlErrorText>
        </FormControlError>
      )}
    </FormControl>
  );
};
