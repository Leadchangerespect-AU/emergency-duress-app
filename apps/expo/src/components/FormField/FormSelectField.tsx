import { useId } from "react";
import { Platform } from "react-native";

import {
  FormControl,
  FormControlError,
  FormControlErrorIcon,
  FormControlErrorText,
  FormControlLabel,
} from "@/components/FormControl";
import Select from "@/components/Select";
import { SelectProps } from "@/components/Select/types";
import { FormFieldProps } from "./types";

export const FormSelectField = ({
  label,
  isRequired,
  isDisabled,
  errors,
  ...props
}: FormFieldProps & SelectProps) => {
  const hasErrors = errors.length > 0;

  const labelId = useId();
  const selectId = useId();
  return (
    <FormControl
      isInvalid={hasErrors}
      isRequired={isRequired}
      isDisabled={isDisabled}
    >
      <FormControlLabel
        id={labelId}
        {...Platform.select({
          web: {
            htmlFor: selectId,
          },
        })}
      >
        {label}
      </FormControlLabel>
      <Select
        {...props}
        aria-labelledby={labelId}
        accessibilityLabel={label}
        id={selectId}
      />
      {hasErrors && (
        <FormControlError>
          <FormControlErrorIcon />
          <FormControlErrorText>{errors.join(", ")}</FormControlErrorText>
        </FormControlError>
      )}
    </FormControl>
  );
};
