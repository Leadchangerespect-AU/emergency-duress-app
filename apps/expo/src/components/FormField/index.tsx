import { FormInputField } from "./FormInputField";
import { FormSelectField } from "./FormSelectField";

export { FormInputField, FormSelectField };
