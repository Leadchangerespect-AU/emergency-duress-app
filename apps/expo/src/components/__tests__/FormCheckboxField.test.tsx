import React from "react";
import { fireEvent, render, screen } from "@testing-library/react-native";

import { FormCheckboxField } from "../FormField/FormCheckboxField";

jest.mock("@expo/vector-icons", () => ({
  Ionicons: "",
}));

describe("FormCheckboxField", () => {
  const defaultProps = {
    value: false,
    onValueChange: jest.fn(),
    errors: [],
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly with default props", () => {
    render(
      <FormCheckboxField {...defaultProps}>Test Checkbox</FormCheckboxField>,
    );
    expect(screen.getByRole("checkbox")).toBeTruthy();
  });

  it("displays error messages", () => {
    const errors = ["This field is required"];
    render(<FormCheckboxField {...defaultProps} errors={errors} />);
    expect(screen.getByText(errors.join(", "))).toBeTruthy();
  });

  it("toggles checkbox state on press", () => {
    const onValueChange = jest.fn();
    render(
      <FormCheckboxField {...defaultProps} onValueChange={onValueChange} />,
    );

    const checkbox = screen.getByRole("checkbox");
    fireEvent.press(checkbox);
    expect(onValueChange).toHaveBeenCalledWith(true);
  });

  it("applies disabled state correctly", () => {
    render(<FormCheckboxField {...defaultProps} disabled />);
    expect(screen.getByRole("checkbox").props.accessibilityState.disabled).toBe(
      true,
    );
  });

  it("applies different colors and sizes", () => {
    render(<FormCheckboxField {...defaultProps} color="secondary" size="lg" />);
    expect(screen.getByRole("checkbox")).toBeTruthy();
  });

  it("matches snapshot", () => {
    render(<FormCheckboxField {...defaultProps} />);
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
