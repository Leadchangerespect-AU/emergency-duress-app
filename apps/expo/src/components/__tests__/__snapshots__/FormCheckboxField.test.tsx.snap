// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheckboxField matches snapshot 1`] = `
<View
  className="my-2"
>
  <View
    className="flex-row undefined"
  >
    <View
      accessibilityRole="checkbox"
      accessibilityState={
        {
          "busy": undefined,
          "checked": false,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      className="mr-2 accent-primary px-2.5 py-2.5"
      collapsable={false}
      focusable={true}
      id=":r5:"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      role="checkbox"
      style={
        [
          {
            "borderColor": "#657786",
            "borderRadius": 2,
            "borderWidth": 2,
            "height": 20,
            "width": 20,
          },
          undefined,
          false,
          {
            "backgroundColor": undefined,
            "borderColor": "#003951",
          },
          undefined,
          false,
        ]
      }
    />
    <Text
      className="text-sm"
    />
  </View>
</View>
`;
